nexus/src/authz/update.rs

```diff
@@ -0,0 +1,457 @@
+    fn get_polar_class_builder() -> oso::ClassBuilder<Self> {
+        oso::Class::builder()
+            .with_equality_check()
+            .add_attribute_getter("version", Self::version)
+    }
+}
+
+/// Describes an updatable, multi-phase host processor image.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct HostImage {
+    version: Version,
+    //phase1, phase2, phase?_tramp, ...
+}
+
+impl HostImage {
+    fn new(version: Version) -> Self {
+        Self { version }
+    }
+
+    fn version(&self) -> Version {
+        self.version
+    }
+}
+
+impl oso::PolarClass for HostImage {
+    fn get_polar_class_builder() -> oso::ClassBuilder<Self> {
+        oso::ClassBuilder::with_constructor(Self::new)
+            .with_equality_check()
+            .add_attribute_getter("version", Self::version)
+    }
+}
+
+/// Describes an updatable Hubris image.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct HubrisImage {
+    version: Version,
+}
+
+impl HubrisImage {
+    fn new(version: Version) -> Self {
+        Self { version }
+    }
+
+    fn version(&self) -> Version {
+        self.version
+    }
+}
+
+impl oso::PolarClass for HubrisImage {
+    fn get_polar_class_builder() -> oso::ClassBuilder<Self> {
+        oso::ClassBuilder::with_constructor(Self::new)
+            .with_equality_check()
+            .add_attribute_getter("version", Self::version)
+    }
+}
+
+/// Describes an update being planned or executed.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Update {
+    pub component: Component,
+    pub image: Image,
+    pub from: Version,
+    pub to: Version,
+}
+
+impl Update {
+    fn new(
+        component: Component,
+        image: Image,
+        from: Version,
+        to: Version,
+    ) -> Self {
+        Self { component, image, from, to }
+    }
+}
+
+impl oso::PolarClass for Update {
+    fn get_polar_class_builder() -> oso::ClassBuilder<Self> {
+        oso::ClassBuilder::with_constructor(Self::new).with_equality_check()
+    }
+}
+
+/// Describes a planned reboot of some component.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Reboot {
+    pub component: Component,
+    pub image: Image,
+    pub to: Version,
+}
+
+impl Reboot {
+    fn new(component: Component, image: Image, to: Version) -> Self {
+        Self { component, image, to }
+    }
+}
+
+impl oso::PolarClass for Reboot {
+    fn get_polar_class_builder() -> oso::ClassBuilder<Self> {
+        oso::ClassBuilder::with_constructor(Self::new).with_equality_check()
+    }
+}
+
+/// Produce a list of plans for updating `component` from version `from`
+/// to `to` that is consistent with the (already loaded into `oso`) Polar
+/// update policy (see `update.polar`). Works by querying the `update`
+/// rule with an unbound `plan` variable and accumulating the results.
+#[allow(dead_code)]
+fn plan_update(
+    oso: &Oso,
+    component: impl ToPolar,
+    from: &Version,
+    to: &Version,
+) -> Result<Vec<PolarValue>> {
```

Comment:
When we chatted about this in the 11/14 update sync, I think there was some hesitation about creating a plan which takes a static view of the system, and makes the call about "how to update the entire rack".

From a versioning point-of-view, this makes total sense, but from a workload and liveness point-of-view, it's somewhat lossy, since a naive linear execution of this plan could run into some problems:
- What if a sled (or component within) fails, or is unplugged, while this plan is executing?
- If workload of a sled is a factor in update ordering, what if the workload changes between executing steps in this plan?

As you raised in that meeting, there was a spectrum of "use polar to enforce policy" (declarative, clear, but strict) vs "write this all in rust code" (ad-hoc, but can incorporate all sorts of information). I wonder if we could take advantage of both, in a system which:

1. Makes a "high-level" decision to attempt to update the rack in manually-written Nexus code
2. Queries the state of the rack through CRDB (to check liveness, versions, and workloads)...
3. ... Uses this planning system to decide if such an update is feasible, and uses it to decide the next *incremental* update (e.g., "update this specific component / sled"). In this case, we could define "incremental" as "step which we can take that we're comfortable will not reduce device availability, for some failure tolerance".
4. Repeats steps (2) and (3) until the planner says "there is no more work to do".

This way, if, for example, sled 5 is unplugged from the rack while sled 2 is updating, we'll re-evaluate the plan on the next loop, and avoid failing the "whole rack update". Additionally, each *incremental* update could be a saga, but the decision to evaluate the "whole rack update" could be considered separately. For example, if we update 90% of the rack, but the updating the last sled fails, we may not actually *want* to rollback. It may be more appropriate to leave the last sled as-is, and retry later, or to evict it from the set of "sleds under active use" by the rack.